esp/esp/accounting/controllers.py

```diff
@@ -519,24 +470,48 @@ def amount_due(self):
         amt_sibling = self.amount_siblingdiscount()
         return amt_request - self.amount_finaid(amt_request, amt_sibling) - amt_sibling - self.amount_paid()
 
-    def clear_payments(self):
-        #   Remove all payments listed for this user at this program
-        line_item_type = self.default_payments_lineitemtype()
-        target_account = self.default_source_account()
-        Transfer.objects.filter(source=None, destination=target_account, user=self.user, line_item=line_item_type).delete()
-
     def submit_payment(self, amount, transaction_id=''):
         #   Create a transfer representing a user's payment for this program
         line_item_type = self.default_payments_lineitemtype()
         target_account = self.default_source_account()
-        return Transfer.objects.create(source=None,
-                                       destination=target_account,
-                                       user=self.user,
-                                       line_item=line_item_type,
-                                       amount_dec=Decimal('%.2f' % amount),
-                                       transaction_id=transaction_id)
+        payment = Transfer.objects.create(source=None,
+                                          destination=target_account,
+                                          user=self.user,
+                                          line_item=line_item_type,
+                                          amount_dec=Decimal('%.2f' % amount),
+                                          transaction_id=transaction_id)
+        self.link_paid_transfers(payment)
+        return payment
+
+    def link_paid_transfers(self, payment):
+        # Given a Transfer representing a payment (e.g. a credit card payment),
+        # find all of the Transfers representing the items that were paid for
+        # and add a link back to the payment.
+
+        # Find the paid transfers by examining Transfers in order of creation
+        # until they sum to the given amount.
+        total = 0
+        target = payment.get_amount()
+        paid_transfers = []
+
+        for transfer in self.get_transfers():
```

Comment:
`get_transfers()` returns a queryset, so you can add additional filters on it. Doing the `paid_in` check here is simple, `is_purchasable` is more complicated* but still doable with Q objects if you want to try that, although it might not be an advantage.

*There are no transfers in any site's database where `line_item` is null. If you want to add a migration to make that an official requirement, it could simplify some things.